### PR TITLE
Fix regex \A and \Z to strictly match string begin/end

### LINKED ISSUE #12278: [BUG] Regression in regexp matching end-of-line
**Describe the bug**
PR https://github.com/rapidsai/cudf/pull/12181 changed the behavior of both `$` and `\Z` to the following:

```
Matches before the final line break in the string (if any) in addition to matching at the very end of the string.
```

This means we can no longer match the very end of the string. For example, `\A\d{4}-\d{2}-\d{2}\Z` will now match `1999-12-29\n`, where it previously did not. There seems to be no way in cuDF to match the end of the string now.

Python has different behavior between `$` and `\Z`, giving the user a choice about how to handle strings ending in newlines. Could we do the same in cuDF?

| Pattern | Meaning |
|--|--|
| `$` | Matches before the final line break in the string (if any) in addition to matching at the very end of the string.
| `\Z` | Matches at the end of the string the regex pattern is applied to.

**Steps/Code to reproduce bug**
See example above.

**Expected behavior**

**Environment overview (please complete the following information)**
N/A

**Environment details**
N/A

**Additional context**
None

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes bug introduced in #12181. The `\A` and `\Z` should strictly match begin/end of the target string regardless of flag settings as already documented https://docs.rapids.ai/api/libcudf/stable/md_regex.html
Also added gtests for these anchor instructions.

Closes #12278 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by andygrove** [comment]
As noted on the issue, this fixes the regression for me. I think the documentation at `cpp/doxygen/regex.md ` also needs to be updated as part of this PR.

File: cpp/tests/strings/contains_tests.cpp [hunk 1]
```diff
@@ -684,6 +684,35 @@ TEST_F(StringsContainsTests, MultiLine)
...
+  auto view = cudf::strings_column_view(input);
+
+  auto pattern = std::string("\\Aabé\\Z");
+  auto prog    = cudf::strings::regex_program::create(pattern);
+  auto prog_ml =
+    cudf::strings::regex_program::create(pattern, cudf::strings::regex_flags::MULTILINE);
+
+  auto results  = cudf::strings::contains_re(view, *prog);
+  auto expected = cudf::test::fixed_width_column_wrapper<bool>({0, 0, 1, 0, 0, 0});
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(*results, expected);
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/tests/strings/contains_tests.cpp:700]: Humnn, it seems that `_EQUIVALENT` is used throughout this test file.

[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/tests/strings/contains_tests.cpp:700]: `EQUIVALENT` is what I want to verify

**Review by ttnghia — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**